sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePartitionedUploader.cs

```diff
@@ -0,0 +1,396 @@
+
+        /// <summary>
+        /// A pool of memory we use to partition the stream into blocks.
+        /// </summary>
+        private readonly ArrayPool<byte> _arrayPool;
+
+        /// <summary>
+        /// The size we use to determine whether to upload as a single PUT BLOB
+        /// request or stage as multiple blocks.
+        /// </summary>
+        private readonly long _singleUploadThreshold;
+
+        /// <summary>
+        /// The size of each staged block.  If null, we'll change between 4MB
+        /// and 8MB depending on the size of the content.
+        /// </summary>
+        private readonly int? _blockSize;
+
+        /// <summary>
+        /// The name of the calling operaiton.
+        /// </summary>
+        private readonly string _operationName;
+
+        public DataLakePartitionedUploader(
+            DataLakeFileClient client,
+            StorageTransferOptions transferOptions,
+            ArrayPool<byte> arrayPool = null,
+            string operationName = null)
+        {
+            _client = client;
+            _arrayPool = arrayPool ?? ArrayPool<byte>.Shared;
+
+            if (transferOptions.MaximumConcurrency.HasValue)
+            {
+                Debug.Assert(transferOptions.MaximumConcurrency > 0);
+                _maxWorkerCount = transferOptions.MaximumConcurrency.Value;
+            }
+            else
+            {
+                _maxWorkerCount = Constants.Blob.Block.DefaultConcurrentTransfersCount;
+            }
+
+            if (transferOptions.MaximumTransferLength.HasValue)
+            {
+                Debug.Assert(transferOptions.MaximumTransferLength > 0);
+                _singleUploadThreshold = Math.Min(transferOptions.MaximumTransferLength.Value, Constants.Blob.Block.MaxUploadBytes);
+            }
+            else
+            {
+                _singleUploadThreshold = Constants.Blob.Block.MaxUploadBytes;
+            }
+
+            _blockSize = null;
+
+            if (transferOptions.MaximumTransferLength != null)
+            {
+                _blockSize = Math.Min(
+                    Constants.DataLake.MaxUploadBytes,
+                    transferOptions.MaximumTransferLength.Value);
+            }
+
+            _operationName = operationName;
+        }
+
+        public async Task<Response<PathInfo>> UploadAsync(
+            Stream content,
+            PathHttpHeaders httpHeaders,
+            DataLakeRequestConditions conditions,
+            IProgress<long> progressHandler,
+            CancellationToken cancellationToken)
+        {
+            await _client.CreateAsync(
+                httpHeaders: httpHeaders,
+                conditions: conditions,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // After the File is Create, Lease ID is the only valid request parameter.
+            conditions = new DataLakeRequestConditions { LeaseId = conditions?.LeaseId };
```

Comment:
Because we actually call Create() as part of DataLakeFileClient.Upload(), we need to clear the non-lease RequestConditions after the initial create.  Otherwise the final Flush() will fail.